### DISCUSSION TITLE
Running Neolink on Mac OS X (10.15)

### DISCUSSION THREAD

**Discussion by michaelrx8** [thirtythreeforty/neolink#253] [Q&A] [answered]
Hi all,
I'm a new Mac user. I recently purchased 2x Reolink Lumus cameras and stumbled upon Neolink. Someone I know was getting rid of an old Mac so I decided to try to get Neolink and Shinobi running on this Mac. Unfortunately, I have underestimated the complexity (or, more to the point, overestimated my talent).

The bug:
I am unable to "run" Neolink. Error received is "cannot execute binary file".

The installation, run and debugging process:
1. Installed Gstreamer (both the runtime and the development installers) using direct link in the instructions (and also tried through Homebrew)
2. Downloaded and unpacked Neolink for macOS x86_64 (from link in instructions)
3. Created configuration file for the cameras
4. Opened Terminal where Neolink was unpacked
5. Tried running command "neolink rtsp --config my_config.toml"; error: "zsh: command not found: neolink"
6. Tried running command "bash neolink rtsp --config my_config.toml"; error: "neolink: neolink: cannot execute binary file"
7. Tried running command "zsh neolink rtsp --config my_config.toml"; error: "neolink:20: parse error near `)'"
8. Used "binutils" to identify Neolink file format; output "mach-o-x86-64"

As mentioned above, I am a totally new Mac user so please forgive my ignorance if there is something I have totally messed up. Any assistance would be much appreciated (BTW, I have tried to resolve this myself through research but I am now close to giving up!)

**Comment by QuantumEntangledAndy** [comment, marked as answer]
Hello, don't panic (HGTG). From your error I feel like this is something simple. Let's some simple stuff.

I'll try explain your error messages:
5. Typo in the command it should be `./neolink rtsp --config my_config.toml` where the `./` means the file is in this current folder. By default programs are only excecuted if they are in the locations `/usr/local/bin` or `/usr/bin` by adding `./` you say don't bother looking for this program in those paces it is here.
6. This is suggesting that either your download is corrupt or you downloaded the wrong version.
7. This error is saying that on line 20 of the neolink code there is an error. First off it should not be reading it line by line as it is not a script. This is suggesting that the download is corrupt.

Can you  run this in the terminal

```
head neolink
```

It will print the first few lines of the file. Post them here

**Reply by michaelrx8** [reply]
Hi Andy,
I thought the answer may have been 42 :)

Anyway, thank you for the prompt reply.
5. Thanks for this! It was really annoying me. I did try to place the "neolink" file in /usr/local/bin but was presented with the same error as I received just now when I tried adding "./"; error: "zsh: permission denied: ./neolink"

Header output:
?????
     ??H__PAGEZEROx__TEXT?0?0__text__TEXT?5p?(?5?__stubs__TEXT??(t???__stub_helper__TEXTd?(|d?(?__gcc_except_tab__TEXT??(8t??(__const__TEXT M*V? M*__unwind_info__TEXTxG0twxG0__eh_frame__TEXT?0??0?__DATA_CONST?0@?0@__got__DATA_CONST?0?0?__const__DATA_CONST?0?0x__DATA4@4@__la_symbol_ptr__DATA4?4?__thread_vars__DATA?4 ?4__data__DATA4h4__thread_data__DATAx
4(x
4__thread_bss__DATA?
4?__bss__DATAx
4?H__LINKEDIT@4?@4?q"?0@4N40@N4@a4PR6?H:?i

                                          P?8?8?A>?B:~ 
                                                       /usr/lib/dyld???3?=?`??0Ey2 

a*(??Q
      pCC/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/libgstrtspserver-1.0.0.dylib
              hCC/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/libgstbase-1.0.0.dylib
                pCC/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/libgstreamer-1.0.0.dylib
                    h??/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/libgobject-2.0.0.dylib
                      h??/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/libglib-2.0.0.dylib
                     h

**Reply by QuantumEntangledAndy** [reply]
Ok it looks like the right file.

Let's do this please.

On the terminal in the same directory as neolink
```bash
chmod +x neolink
./neolink --help
```

The `chmod +x` give it execute permissions

**Reply by michaelrx8** [reply]
Ok! That seems to have been the issue. Resolved now (thank you very much!). I can access the help "menu".

I also tried to run "./neolink rtsp --config config.toml" and received an error which I was then able to fix (something with the way the config file was saved; some "garbage" text at the beginning).

Ran the command again and ...
[2022-01-30T10:56:51Z INFO  neolink] Neolink 57c55e4121c0fe7abf9916f00df8d4bb9dd468c8 release
[2022-01-30T10:56:54Z INFO  neolink::rtsp] front: Connecting to camera at Address: 192.168.1.108:9000
[2022-01-30T10:56:54Z INFO  neolink::rtsp] front: Logging in
[2022-01-30T10:56:54Z INFO  neolink::rtsp] front: Connecting to camera at Address: 192.168.1.108:9000
[2022-01-30T10:56:54Z INFO  neolink::rtsp] front: Logging in
[2022-01-30T10:56:54Z ERROR neolink_core::bc_protocol::connection::bcconn] Deserialization error: Deserialization error
[2022-01-30T10:56:54Z ERROR neolink_core::bc_protocol::connection::bcconn] caused by: Parsing error
[2022-01-30T10:56:54Z INFO  neolink::rtsp] front: Connected and logged in
[2022-01-30T10:56:54Z INFO  neolink::rtsp] front: Starting video stream Sub Stream (Fluent)
[2022-01-30T10:56:55Z INFO  neolink::rtsp] front: Connected and logged in
[2022-01-30T10:56:55Z INFO  neolink::rtsp] front: Camera time is already set: 2022-01-30 21:57:07 +10
[2022-01-30T10:56:55Z INFO  neolink::rtsp] front: Camera reports firmware version v2.0.0.687_20102800
[2022-01-30T10:56:55Z INFO  neolink::rtsp] front: Starting video stream Main Stream (Clear)

Looks like it is working now (I'll have to figure out what that error is).
Again, thank you so much for your help. You have reignited my drive to get this up and running.

**Reply by QuantumEntangledAndy** [reply]
Yes just missing exec permissions.

When you ran it as `bash ...` and `zsh ...` it didn't know it was an exec so it just tried to read it as a bash or zsh script.

**Reply by michaelrx8** [reply]
Ah! Got it. I saw a few of your posts where you assumed that people knew how to run the app in terminal. I was not (but now am) one of those people. Thank you.
